src/Datadog.Trace/RuntimeMetrics/RuntimeMetricsWriter.cs

```diff
@@ -0,0 +1,131 @@
+using System;
+using System.Diagnostics;
+using System.Runtime.ExceptionServices;
+using System.Threading;
+using Datadog.Trace.Vendors.StatsdClient;
+
+namespace Datadog.Trace.RuntimeMetrics
+{
+    internal class RuntimeMetricsWriter : IDisposable
+    {
+        private static readonly string[] GcCountMetricNames = { "runtime.dotnet.gc.count.gen0", "runtime.dotnet.gc.count.gen1", "runtime.dotnet.gc.count.gen2" };
+        private static readonly string[] CompactingGcTags = { "compacting_gc:true" };
+        private static readonly string[] NotCompactingGcTags = { "compacting_gc:false" };
+
+        private readonly int _delay;
+
+        private readonly IDogStatsd _statsd;
+        private readonly Timer _timer;
+        private readonly Process _currentProcess;
+
+#if NETCOREAPP
+        private readonly RuntimeEventListener _listener;
+#endif
+
+        private readonly Timing _contentionTime = new Timing();
+
+        private TimeSpan _previousUserCpu;
+        private TimeSpan _previousSystemCpu;
+
+        private long _contentionCount;
+
+        public RuntimeMetricsWriter(IDogStatsd statsd, int delay)
+        {
+            _delay = delay;
+            _statsd = statsd;
+            _timer = new Timer(_ => PushEvents(), null, delay, delay);
+            _currentProcess = Process.GetCurrentProcess();
```

Comment:
Maybe not needed now, but if we end up running this on .NET Framework and have no permissions except execution permissions, we'll throw when this method (and any method accessing the `Process` type) is JIT compiled because we'll fail the LinkDemand of FullTrust. The way I worked around this in #1001 was to have helper methods deal with the `Process` accesses

Reply:
Done